universal-application-tool-0.0.1/app/repository/VersionRepository.java

```diff
@@ -0,0 +1,328 @@
+                      .where()
+                      .raw("active.name = name")
+                      .eq("lifecycle_stage", LifecycleStage.DRAFT)
+                      .query())
+              .findList()) {
+        uneditedActive.setLifecycleStage(LifecycleStage.OBSOLETE);
+        uneditedActive.save();
+        Program newVersion = new Program(uneditedActive.getProgramDefinition());
+        newVersion.setVersion(version);
+        newVersion.setLifecycleStage(LifecycleStage.ACTIVE);
+        this.programRepository.insertProgramSync(newVersion);
+      }
+
+      // Then, for each program, if a draft exists, publish the draft.
+      ebeanServer
+          .find(Program.class)
+          .where()
+          .eq("lifecycle_stage", LifecycleStage.DRAFT)
+          .findList()
+          .stream()
+          .forEach(
+              draft -> {
+                draft.setVersion(version);
+                this.publishProgram(draft, transaction);
+              });
+      ebeanServer.commitTransaction();
+    } finally {
+      ebeanServer.endTransaction();
+    }
+  }
+
+  public long getNextVersion() {
+    Optional<Program> draft =
+        ebeanServer
+            .find(Program.class)
+            .where()
+            .eq("lifecycle_stage", LifecycleStage.DRAFT)
+            .isNotNull("version")
+            .setMaxRows(1)
+            .findOneOrEmpty();
+    if (draft.isPresent()) {
+      return draft.get().getVersion();
+    }
+    // If there are no drafts, add one to any active program
+    Optional<Program> active =
+        ebeanServer
+            .find(Program.class)
+            .where()
+            .eq("lifecycle_stage", LifecycleStage.ACTIVE)
+            .isNotNull("version")
+            .setMaxRows(1)
+            .findOneOrEmpty();
+    if (active.isPresent()) {
+      return active.get().getVersion() + 1;
+    }
+
+    // If there are no programs, we are pretty sure the correct version is 1, but
+    // we might also be in a questions-only test.  Search questions just in case.
+    Optional<Question> draftQuestion =
+        ebeanServer
+            .find(Question.class)
+            .where()
+            .eq("lifecycle_stage", LifecycleStage.DRAFT)
+            .isNotNull("version")
+            .setMaxRows(1)
+            .findOneOrEmpty();
+    if (draftQuestion.isPresent()) {
+      return draftQuestion.get().getVersion();
+    }
+    // If there are no drafts, add one to any active program
+    Optional<Question> activeQuestion =
+        ebeanServer
+            .find(Question.class)
+            .where()
+            .eq("lifecycle_stage", LifecycleStage.ACTIVE)
+            .isNotNull("version")
+            .setMaxRows(1)
+            .findOneOrEmpty();
+    if (activeQuestion.isPresent()) {
+      return activeQuestion.get().getVersion() + 1;
+    }
+
+    return 1L;
+  }
+
+  private Optional<Question> getLatestVersionOfQuestion(Transaction transaction, long questionId) {
```

Comment:
Suggestion (non-blocking): it'd be good to make clear somewhere in documentation and/or comments that questions are linked via their immutable name.

Reply:
Commented in the model and service classes.